actix-multipart/src/form/mod.rs

```diff
@@ -0,0 +1,832 @@
+    /// to ensure that the form limits are not exceeded.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` - The number of bytes being read from this chunk
+    /// * `in_memory` - Whether to consume from the memory limits
+    pub fn try_consume_limits(
+        &mut self,
+        bytes: usize,
+        in_memory: bool,
+    ) -> Result<(), MultipartError> {
+        self.total_limit_remaining = self
+            .total_limit_remaining
+            .checked_sub(bytes)
+            .ok_or(MultipartError::Payload(PayloadError::Overflow))?;
+        if in_memory {
+            self.memory_limit_remaining = self
+                .memory_limit_remaining
+                .checked_sub(bytes)
+                .ok_or(MultipartError::Payload(PayloadError::Overflow))?;
+        }
+        if let Some(field_limit) = self.field_limit_remaining {
+            self.field_limit_remaining = Some(
+                field_limit
+                    .checked_sub(bytes)
+                    .ok_or(MultipartError::Payload(PayloadError::Overflow))?,
+            );
+        }
+        Ok(())
+    }
+}
+
+/// Typed `multipart/form-data` extractor.
+///
+/// To extract typed data from a multipart stream, the inner type `T` must implement the
+/// [`MultipartFormTrait`] trait, you should use the [`macro@MultipartForm`] macro to derive this for
+/// your struct.
+///
+/// Use [`MultipartFormConfig`] to configure extraction options.
+#[derive(Deref, DerefMut)]
+pub struct MultipartForm<T: MultipartFormTrait>(pub T);
+
+impl<T: MultipartFormTrait> MultipartForm<T> {
+    /// Unwrap into inner `T` value.
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T> FromRequest for MultipartForm<T>
+where
+    T: MultipartFormTrait,
+{
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
+
+    #[inline]
+    fn from_request(req: &HttpRequest, payload: &mut Payload) -> Self::Future {
+        let mut payload = Multipart::new(req.headers(), payload.take());
+        let config = MultipartFormConfig::from_req(req);
+        let mut limits = Limits::new(config.total_limit, config.memory_limit);
+        let req = req.clone();
+        let req2 = req.clone();
+        let err_handler = config.err_handler.clone();
+
+        async move {
+            let mut state = State::default();
+            // We need to ensure field limits are shared for all instances of this field name
+            let mut field_limits = HashMap::<String, Option<usize>>::new();
+
+            while let Some(field) = payload.try_next().await? {
+                // Retrieve the limit for this field
+                let entry = field_limits
+                    .entry(field.name().to_owned())
+                    .or_insert_with(|| T::limit(field.name()));
+                limits.field_limit_remaining = entry.to_owned();
+
+                T::handle_field(&req, field, &mut limits, &mut state).await?;
```

Comment:
I realize this might be difficult with the &mut limits and &mut state in here, but would there be any way to race `handle_field` against `payload.try_next()`

In my crate, actix-form-data, I have a FuturesUnordered that I push the field handler futures into. This allows field handlers to drop their Field stream upon completion and make the remaining work concurrent with the next field handlers

Reply:
hmmm... that is interesting I never even thought about that. I think to do that we would have to synchronise on the limits & state map, which is do-able.

I do wonder though if it actually brings any noticable performance improvement - since the multipart as a whole is being received as a single stream, presumably the next part can't be received until all of the previous part has been buffered - I guess it would depend on how much buffering actix-multipart does internally?

Reply:
if a handler does something like
```rust
let bytes = Vec::new();
while let Some(res) = field.next().await {
    bytes.extend_from_slice(&res?);
}
drop(field);
let result =  send_these_bytes_somewhere(bytes).await;
```
then the `send_these_bytes_somewhere` could be run concurrently with the next field handler, since the field was completely read and dropped before that is executed.
 
I use this in my application pict-rs to better allow concurrently uploading files to object storage: https://git.asonix.dog/asonix/pict-rs/src/branch/main/src/store/object_store.rs#L246
This allows making additional requests to object storage without waiting for the existing in-flight requests to complete

Reply:
oh I see - you're effectively adding your own internal buffer to the handler

Reply:
does this mean your server memory usage is only bounded by the overall form limit, e.g. if `send_these_bytes_somewhere` / your object storage server is running quite slowly, whilst the multipart upload is very fast, then nearly all of the contents could be be read into memory before getting uploaded?

Reply:
Yes, that is true. actix-form-data has per-field size limits and total field count limits, so the maximum possible memory usage is `max_field_size * max_field_count`, and in theory that limit could be hit by slow object storage.

Reply:
I've done some thinking on this and I think I can get away with using `spawn` to make this concurrent without changing the design here.